ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java

```diff
@@ -298,7 +308,7 @@ public EntryWithData getEntryWithData(long index) throws RaftLogIOException {
     try {
       CompletableFuture<ByteString> future = null;
       if (stateMachine != null) {
-        future = stateMachine.data().read(entry).exceptionally(ex -> {
+        future = stateMachine.data().read(entry, server.getTransactionContext(entry, false)).exceptionally(ex -> {
```

Comment:
The `TransactionManager` manages the `TransactionContext` for non-leaders, while `PendingRequests` manages the context for the leader, am I getting it right?
Normally when `DataApi.read` is called, this peer is leader and is going to replicate the full log to non-leaders. Why do we need to call `getTransactionContext` here?

Reply:
You are right that the context should be obtained from LeaderState.  Let me update the change.